### ISSUE TITLE
log4j using by the agent will attempt to loading the application's log4j2 class, leads to  ClassNotFoundException

### ISSUE THREAD

**Issue by CoderLan0668** [elastic/apm-agent-java#1316]
the log here 
```
ERROR StatusLogger Unable to create custom ContextSelector. Falling back to default.
 java.lang.ClassNotFoundException: org/apache/logging/log4j/core/async/AsyncLoggerContextSelector
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.loadClass(LoaderUtil.java:168)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.newInstanceOf(LoaderUtil.java:207)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.newCheckedInstanceOf(LoaderUtil.java:228)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.newCheckedInstanceOfProperty(LoaderUtil.java:253)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.util.Loader.newCheckedInstanceOfProperty(Loader.java:338)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.Log4jContextFactory.createContextSelector(Log4jContextFactory.java:98)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.Log4jContextFactory.<init>(Log4jContextFactory.java:59)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.LogManager.<clinit>(LogManager.java:94)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.spi.ThreadContextMapFactory.createThreadContextMap(ThreadContextMapFactory.java:99)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.ThreadContext.init(ThreadContext.java:225)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.ThreadContext.<clinit>(ThreadContext.java:203)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.ContextDataInjectorFactory.createDefaultInjector(ContextDataInjectorFactory.java:83)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.ContextDataInjectorFactory.createInjector(ContextDataInjectorFactory.java:67)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.ContextMapLookup.<init>(ContextMapLookup.java:34)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.Interpolator.<init>(Interpolator.java:120)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.Interpolator.<init>(Interpolator.java:87)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory.<init>(ConfigurationFactory.java:133)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory$Factory.<init>(ConfigurationFactory.java:354)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory$Factory.<init>(ConfigurationFactory.java:354)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory.<clinit>(ConfigurationFactory.java:131)
	at co.elastic.apm.agent.logging.LoggingConfiguration.init(LoggingConfiguration.java:198)
	at co.elastic.apm.agent.impl.ElasticApmTracerBuilder.<init>(ElasticApmTracerBuilder.java:92)
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:127)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:100)
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:54)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
```

**Comment by eyalkoren** [comment]
Thanks for reporting!
Which agent version are you using?
This error seems to be handled by using a default context selector instead. Besides this log message, is there anything not working as expected?

**Comment by felixbarny** [comment]
I've also noticed an issue with log4j and indy plugins when testing #893 with this application: https://github.com/micrometer-metrics/micrometer-samples-spring-boot. It might be related.

When the logger gets initialized I get this message:

``` 
2020-07-29 17:50:48,312 localhost-startStop-1 ERROR No Log4j 2 configuration file found. Using default configuration (logging only errors to the console), or user programmatically provided configurations. Set system property 'log4j2.debug' to show Log4j 2 internal initialization logging. See https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions on how to configure Log4j 2
```

And no logs from the indy plugin appear in the log file.

**Comment by CoderLan0668** [comment]
> Thanks for reporting!
> Which agent version are you using?
> This error seems to be handled by using a default context selector instead. Besides this log message, is there anything not working as expected?

the agent not working, No Class has been retransformed. the agent was built from master branch, lastest comit is `518afc790674f311b00fb188fe2444b1a6a81a6a`

**Comment by CoderLan0668** [comment]
and get the same message like  @felixbarny   got

**Comment by eyalkoren** [comment]
> and get the same message like @felixbarny got

I verified that the agent works properly when only this error message is written. 
@CoderLan0668 I think what's different in your environment is that there are non-default log4j2 configurations on the classpath. If that is indeed the case, please provide you log4j config file.

I still can't see how this prevents the agent from working. Is the application or application-logging affected by this as well? 

I will look for the best way to eliminate this shading problem. In the mean time, please also provide your agent setup/configuration and a `DEBUG` log of the agent (setting `log_level=DEBUG`).

**Comment by felixbarny** [comment]
We could instrument `org.apache.logging.log4j.util.PropertiesUtil.Environment#get` and always return null. This makes sure our log4j2 never reads any system properties.

**Comment by eyalkoren** [comment]
That's true, and the system will fall back to the defaults. 
However we will still have the error mentioned in https://github.com/elastic/apm-agent-java/issues/1316#issuecomment-665750106.
If we assume that falling back to the defaults is what we want, shutting off the `StatusLogger` for the initialization time should be a valid choice that will eliminate this error message as well.
I'll try it out.

**Comment by felixbarny** [comment]
The `ERROR No Log4j 2 configuration file found` error should be fixed by #1320

**Comment by eyalkoren** [comment]
> The ERROR No Log4j 2 configuration file found error should be fixed by #1320

Huh, I can't see how but good to know anyway 😄 !

@CoderLan0668 can you try [the artifact of this PR](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp/detail/PR-1320/1/artifacts) and see if it fixes the issue you see as well?

If not, I will add something to make sure our log4j doesn't try to load classes it shouldn't.
Your answers to my questions above can also help.

**Comment by CoderLan0668** [comment]
@eyalkoren 
I‘m sorry，the agent not working can not be reproduced anymore， the agent can work.

trying  the snapshot version you provided， i got another error log：

`
ERROR StatusLogger Unable to create custom ContextSelector. Falling back to default.
 java.lang.ClassCastException: Cannot cast org.apache.logging.log4j.core.async.AsyncLoggerContextSelector to co.elastic.apm.agent.shaded.apache.logging.log4j.core.selector.ContextSelector
	at java.lang.Class.cast(Class.java:3369)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.newCheckedInstanceOf(LoaderUtil.java:228)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.util.LoaderUtil.newCheckedInstanceOfProperty(LoaderUtil.java:253)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.util.Loader.newCheckedInstanceOfProperty(Loader.java:338)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.Log4jContextFactory.createContextSelector(Log4jContextFactory.java:98)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.Log4jContextFactory.<init>(Log4jContextFactory.java:59)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.LogManager.<clinit>(LogManager.java:94)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.spi.ThreadContextMapFactory.createThreadContextMap(ThreadContextMapFactory.java:99)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.ThreadContext.init(ThreadContext.java:225)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.ThreadContext.<clinit>(ThreadContext.java:203)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.ContextDataInjectorFactory.createDefaultInjector(ContextDataInjectorFactory.java:83)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.impl.ContextDataInjectorFactory.createInjector(ContextDataInjectorFactory.java:67)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.ContextMapLookup.<init>(ContextMapLookup.java:34)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.Interpolator.<init>(Interpolator.java:120)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.lookup.Interpolator.<init>(Interpolator.java:87)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory.<init>(ConfigurationFactory.java:133)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory$Factory.<init>(ConfigurationFactory.java:354)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory$Factory.<init>(ConfigurationFactory.java:354)
	at co.elastic.apm.agent.shaded.apache.logging.log4j.core.config.ConfigurationFactory.<clinit>(ConfigurationFactory.java:131)
	at co.elastic.apm.agent.logging.LoggingConfiguration.init(LoggingConfiguration.java:198)
	at co.elastic.apm.agent.impl.ElasticApmTracerBuilder.<init>(ElasticApmTracerBuilder.java:92)
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:103)
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
`

As the log above， i add a param `-DLog4jContextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector`， but the agent load it.

**Comment by eyalkoren** [comment]
Instrumentation of the shaded log4j lib would have been an elegant solution, however the Logging initialization takes place before instrumentations are applied and changing that seems like an overkill for this specific issue.

Shading for the "problematic" system properties is a good idea as well, however, a quick search showed there are multiple such properties, located in different classes.

I eventually chose to make the log4j `StatusLogger` "silence" only during initialization. The assumption is that as long as such errors are handled, log4j will be functional using the default settings (which is the required state). Non handled exceptions during initialization will be printed to system err.

**Comment by eyalkoren** [comment]
@CoderLan0668 Please try [this bugfix](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp/detail/PR-1322/1/artifacts) and see if the error is gone and everything works as expected.

**Comment by CoderLan0668** [comment]
@eyalkoren  The error logs has gone, and the agent work well.

**Comment by felixbarny** [comment]
> Shading for the "problematic" system properties is a good idea as well, however, a quick search showed there are multiple such properties, located in different classes, without a strict common prefix or pattern, so might be easy to miss stuff.

Don't all or at least most properties contain `log4j`?
Could we shade occurrences of `log4j` to `elasticapmlog4j`? We could also use `<includes>` and `<excludes>` (https://maven.apache.org/plugins/maven-shade-plugin/examples/class-relocation.html) to limit that to only the log4j classes.


Reopening as we didn't merge a fix yet.

**Comment by eyalkoren** [comment]
> Don't all or at least most properties contain log4j?
Could we shade occurrences of log4j to elasticapmlog4j? 

I am pretty sure shading occurrence of `log4j` in the log4j library will capture much more than we intend to capture, which may or may not have side effects.
As said above, I think silencing the `StatusLogger` only for the initialization stage and only for handled exceptions is safe enough. Unless you strongly disagree, I rather merge it this way than spend more time on something that only affects logs.

**Comment by felixbarny** [comment]
I'm still a bit skeptical about whether it works but we can try silencing the logger. If it turns out to be problematic, we can always change it later on. The thing I'm worried about is that we share the same global system properties with the application. If someone sets these properties they will also change the behavior for the logger in the agent. But it may be a non-issue in practice.

**Comment by eyalkoren** [comment]
> The thing I'm worried about is that we share the same global system properties with the application. If someone sets these properties they will also change the behavior for the logger in the agent. 

Absolutely agree, I was also worried about that. That's why I loved the idea of instrumenting `org.apache.logging.log4j.util.PropertiesUtil#getStringProperty(java.lang.String)` to always return `null`, as it is used by all config types, not only Strings. I thought of applying only this single instrumentation first before log4j initialization, but this seems like an overkill.
However, the more I looked into this, I realized that most of the configurations we really care about (rollover decisions, file patterns, layouts etc.) are set programmatically, so not sure it will have a real effect.
Anyway, as you say, if we find out it is not good enough, we can reiterate.
